cpp/include/raft/core/mdarray.hpp

```diff
@@ -57,6 +87,117 @@ template <typename ElementType,
+ */
+template <typename T>
+inline constexpr bool is_mdspan_v =
+  std::conjunction_v<__is_mdspan_t<T>,
+                     std::disjunction<__is_device_mdspan<T, true>, __is_host_mdspan<T, true>>>;
+
+/**
+ * @brief Interface to implement an owning multi-dimensional array
+ *
+ * raft::array_interace is an interface to owning container types for mdspan.
+ * Check implementation of raft::mdarray which implements raft::array_interface.
+ * This interface provides virtual method `view()` which is to be overridden by
+ * the implementing class. `view()` must return an object of type raft::host_mdspan
+ * or raft::device_mdspan or any types derived from the them.
+ */
+template <typename ElementType, typename Extents, typename LayoutPolicy, typename ContainerPolicy>
+class array_interface {
+  static_assert(!std::is_const<ElementType>::value,
+                "Element type for container must not be const.");
+
+ public:
+  using extents_type = Extents;
+  using layout_type  = LayoutPolicy;
+  using element_type = ElementType;
+
+  // Naming: ref impl: container_policy_type, proposal: container_policy
+  using container_policy_type = ContainerPolicy;
+  using container_type        = typename container_policy_type::container_type;
+
+ private:
+  template <typename E,
+            typename ViewAccessorPolicy =
+              std::conditional_t<std::is_const_v<E>,
+                                 typename container_policy_type::const_accessor_policy,
+                                 typename container_policy_type::accessor_policy>>
+  using view_type_impl =
+    std::conditional_t<container_policy_type::is_host_type::value,
+                       host_mdspan<E, extents_type, layout_type, ViewAccessorPolicy>,
+                       device_mdspan<E, extents_type, layout_type, ViewAccessorPolicy>>;
+
+ public:
+  /**
+   * \brief the mdspan type returned by view method.
+   */
+  using view_type       = view_type_impl<element_type>;
+  using const_view_type = view_type_impl<element_type const>;
+
+  /**
+   * @brief Get a mdspan that can be passed down to CUDA kernels.
+   */
+  virtual view_type view() noexcept = 0;
+  /**
+   * @brief Get a mdspan that can be passed down to CUDA kernels.
+   */
+  virtual const_view_type view() const noexcept = 0;
+};
+
+template <typename ElementType, typename Extents, typename LayoutPolicy, typename ContainerPolicy>
+void __takes_an_array_interface_ptr(
+  array_interface<ElementType, Extents, LayoutPolicy, ContainerPolicy>*);
+
+template <typename T, typename = void>
+struct __is_array_interface : std::false_type {
+};
+
+template <typename T>
+struct __is_array_interface<
+  T,
+  std::void_t<decltype(__takes_an_array_interface_ptr(std::declval<T*>()))>> : std::true_type {
+};
+
+/**
+ * @\brief Boolean to determine if template type T is raft::array_interface or derived type
+ */
+template <typename T>
+inline constexpr bool is_array_interface_v = __is_array_interface<std::remove_const_t<T>>::value;
```

Comment:
I propose that we give users two options here:
1. They straightforward implement `array_interface`
2. For users with existing owning types, we just ask that they add a `view()` method that returns a host_mdspan or device_mdspan or a derived type

Point 1 is already implemented, and point 2 allows users to update their existing their code minimally.